### ISSUE TITLE
Allow async function to return union of `Promise`s

### ISSUE THREAD

**Issue by sterliakov** [microsoft/TypeScript#51293]
# Suggestion

I think that the code below should be considered valid. At the very least, it is more semantically valid: you say "return either one promise or another", but not "return promise resolving to one of the types". For me class `Y` looks much better, but currently tsc declines it with message

> The return type of an async function or method must be the global Promise<T> type. Did you mean to write 'Promise<string | number>'?

Snippet is copied below, and here's [Workbench Repro](https://www.typescriptlang.org/dev/bug-workbench/?#code/MYGwhgzhAEAa0G8CwAoa1IE8B2xoDMAKADwC5oAXAJwFcBTASnIAUqB7AWwEsI6AebDQ4AjOlQB8AblTosuAiXL4wIXk2itOPfhGpdsAcykyMEHHiJlowtmxB0w2dZu68BQ0VWgAfaLqr6RtJopuYKVjZ2Dk4s7K78giJiPn56huKIJuhUdBQ0VNjQxNAA-NAAjNDkAETl1cHoAL6ozSiooJAwAJqZIXIWipS0jLFabomexn1m8pZKKmqj8Xz+gVOyMwMRtvaOznHa7kkSKS6Hq+kNobODkbsxGgfjHmIZvmduF0a96Nm5+YVimVKjU6ldWo0gA) 

Also, it is absolutely not clear from documentation whether union and intersection types are associative with generics (e.g. is `G<A | B>` equivalent to `G<A> | G<B>`? From type theory point of view - definitely not, but typescript may have decided to treat them as equal. If they are not, then class `Y` is the **only** correct solution, and `X` is false-negative pass.

## 🔍 Search Terms

Promise, overload, union, union-type, type-union, async, async-function, associativity

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

I wish `async` functions could return union of `Promise`'s

## 📃 Motivating Example

This feature is of great importance for overloaded async functions. In addition to type theory adherence, it allows to declare `Promise<something>` types in advance for reusability, and then use unions of pre-defined types. The latter is not this important, because it is OK to define type `something` before and wrap with Promise only at function definition level, but it's less clear and more verbose IMO.

## 💻 Use Cases

```ts repro
class X {
  async f(x: true): Promise<number>;
  async f(x: false): Promise<string>;
  async f(x: boolean): Promise<number | string>;
  async f(x: boolean): Promise<number | string> {
    return x ? 1 : "1";
  }
}

class Y {
  async f(x: true): Promise<number>;
  async f(x: false): Promise<string>;
  async f(x: boolean): Promise<number> | Promise<string>;
  async f(x: boolean): Promise<number> | Promise<string> {
    return x ? 1 : "1";
  }
}
```

**Comment by fatcerberus** [comment]
`G<T | U>` is not equivalent to `G<T> | G<U>` *in general*, but in the specific case of `Promise`, they are at least isomorphic if not equivalent: a promise yields only one value, and you can’t tell `T` from `U` until it does.  So it doesn’t really matter where you put the quantifier.

**Comment by sterliakov** [comment]
I agree that it may be not very important, but (just learning TypeScript for 3-4 months with very unsatisfying experience, coming from python typing system plus Rust static typing and related academic background) for me it's just a nonsense to fight compiler creating types which do not mean what I want. Is it hard to implement? I thought this rule is pretty self-contained ("async functions must return Promise subtype"), and it should be not very difficult to expand it to "union of Promises". I'm ready to work on this feature, if there is any interest (and I don't throw typescript away earlier). 

Union (join) of promises is intuitively not equivalent to promise of union, and opposite compiler preference drives me mad. I help to maintain one ts project, and one of the classes has approx. 70 overloaded async functions. They'd be just a single Protocol + decorator in python, but remain four lines each and anti-DRY in TypeScript. Looking at them, I feel really disappointed, because it looks like a mess of hacks and nothing more.

My question primary is "if it is a design decision, what is the cause of it? If it was just simpler to implement, how difficult would it be to extend allowed return types?"

One more reasoning I discovered after posting is related to using multiple promise kinds: for example, there is some user-defined Promise alternative `UserPromise`, that encapsulates some logic and provides similar interface. 

Some function may return `Promise<string>` on failure and `UserPromise<number>` that resolves to, say, `{number: number}` (thanks to wrapper logic, similar to `.then(number => ({number}))`) - and whole return type is `Promise<string> | UserPromise<number>`, while union approach results in `Promise<string | {number: number}>`, which violates DRY and makes me cry.

**Comment by fatcerberus** [comment]
For what it’s worth `async` functions *can’t* return custom promises.  They will be re-wrapped and the caller will only see standard promises.  The declared return type must match what the caller sees.

**Comment by fatcerberus** [comment]
You seem to be coming from a background of much more consistent type systems where things like mathematical soundness matter.  That’s not TypeScript’s goal - TS intends to model idiomatic JS with its type system whenever possible, and this leads to a lot of inconsistencies in the type system that make no sense from a type theoretical perspective.  I fear you may never be happy with TS on its own terms in this case ☹️

**Comment by sterliakov** [comment]
> For what it’s worth `async` functions _can’t_ return custom promises. They will be re-wrapped and the caller will only see standard promises. The declared return type must match what the caller sees.

Wow, I thought it's possible, but custom `Promise` subclass is really wrapped with a regular `Promise`, thank you for sharing! It may be important.

**Comment by MartinJohns** [comment]
Welcome to the world of JavaScript.

**Comment by RyanCavanaugh** [comment]
The proposed reduction rule is valid today, but I'm not sure what to make of it.

There's really nothing stopping TC39 from, tomorrow, deciding that Promises should have a `.thenAgain` method that runs a fresh invocation of the callback, and at that point the rule that `Promise<T | U> === Promise<T> | Promise<U>` stops being true. Then we'd be in quite a pickle.

The upside is what? You haven't been stopped from writing a valid spelling of a type which you agree to be equivalent.

I don't find the preference described in the OP particularly compelling either, to be honest. The desugaring of this function is
```ts
class Y {
    f(x) {
        return new Promise(resolve => {
            resolve(x ? 1 : "1");
        });
    }
}
```
There's only one `return` statement here, so it doesn't seem correct to say that there are two possible output candidates, and the program is only valid if `Promise` provides the type `(value: string | number) => unknown` to `resolve`. So why isn't this a `Promise<string | number>` ?

This particular rule isn't something we've heard other feedback on despite obviously ~everyone interacting with Promises in one way or another, and there are no type checkers which don't reject at least some valid programs, so on net I don't see an action item here.

**Comment by sterliakov** [comment]
@RyanCavanaugh  I can write the same as

```
class Y {
    f(x) {
        if (x)
            return new Promise(resolve => resolve(1));
        else
            return new Promise(resolve => resolve('1'));
    }
}
```

And here natural type is union of promises. It's just a toy implementation. If real-world example is interesting, this request arose related to a library with legacy interface that must remain for backwards compatibility. Almost all methods accept `callback` as last argument, and return `Promise<void>` if it is given and not-null and `Promise<SomeReturnType>` otherwise. This results in overloads everywhere, and I describe the return type "as one promise or another", but not "promise of void or SomeReturnType" - since I cannot just create a nominal interface and use it as callable signature, it annoys me. 

Just to note: I don't mind closing this issue, because it does not seem to have enough community support, and I'm far from experienced/motivated TypeScript user. I just encountered this quite non-obvious compiler claim and raised an issue to let active community members decide. 

The fact that `Promise<T | U> ~ (Promise<T> | Promise<U>)` does not seem to be documented somewhere, and (as answered above) the same luckily does not hold for arbitrary generics. However, IMO the error message can be more clear, because - again - the fact above is really counter-intuitive.

**Comment by fatcerberus** [comment]
It’s not directly documented because it’s just a side-effect of [structural typing](https://www.typescriptlang.org/docs/handbook/type-compatibility.html).  There’s no special carve-out for `Promise` in this regard.

Unlike in nominal type systems, whether union-of-generic is equivalent to generic-of-union in TS depends on the makeup of the type involved.
